blocks/functions.js

```diff
@@ -44,44 +44,931 @@ Blockly.PXTBlockly.FunctionUtils = {};
+  this.addParam_('string', Blockly.Msg.FUNCTIONS_DEFAULT_STRING_ARG_NAME);
+};
+
+/**
+ * Externally-visible function to add a number argument to the function
+ * declaration.
+ * @public
+ */
+Blockly.PXTBlockly.FunctionUtils.addNumberExternal = function() {
+  this.addParam_('number', Blockly.Msg.FUNCTIONS_DEFAULT_NUMBER_ARG_NAME);
+};
+
+/**
+ * Externally-visible function to add a custom argument to the function
+ * declaration.
+ * @param {string} typeName The custom type of the param, as would be emitted
+ *  to Typescript.
+ * @public
+ */
+Blockly.PXTBlockly.FunctionUtils.addCustomExternal = function(typeName) {
+  this.addParam_(typeName, Blockly.Msg.FUNCTIONS_DEFAULT_CUSTOM_ARG_NAME);
+};
+
+/**
+ * Callback to remove a field, only for the declaration block.
+ * @param {Blockly.Field} field The field being removed.
+ * @public
+ */
+Blockly.PXTBlockly.FunctionUtils.removeFieldCallback = function(field) {
+  var inputNameToRemove = null;
+  for (var n = 0; n < this.inputList.length; n++) {
+    if (inputNameToRemove) {
+      break;
+    }
+    var input = this.inputList[n];
+    if (input.connection) {
+      var target = input.connection.targetBlock();
+      if (!target) {
+        continue;
+      }
+      if (target.getField(field.name) === field) {
+        inputNameToRemove = input.name;
+      }
+    } else {
+      for (var j = 0; j < input.fieldRow.length; j++) {
+        if (input.fieldRow[j] == field) {
+          inputNameToRemove = input.name;
+        }
+      }
+    }
+  }
+  if (inputNameToRemove) {
+    Blockly.WidgetDiv.hide(true);
+    this.removeInput(inputNameToRemove);
+    this.updateFunctionSignature();
+    this.updateDisplay_();
+  }
+};
+
+/**
+ * Callback to pass removeField up to the declaration block from arguments.
+ * @param {Blockly.Field} field The field being removed.
+ * @public
+ */
+Blockly.PXTBlockly.FunctionUtils.removeArgumentCallback_ = function(field) {
+  if (this.parentBlock_ && this.parentBlock_.removeFieldCallback) {
+    this.parentBlock_.removeFieldCallback(field);
+  }
+};
+
+/**
+ * Function calls cannot exist without the corresponding function definition.
+ * Enforce this link whenever an event is fired.
+ * @param {!Blockly.Events.Abstract} event Change event.
+ * @this Blockly.Block
+ */
+Blockly.PXTBlockly.FunctionUtils.onCallerChange = function(event) {
+  if (!this.workspace || this.workspace.isFlyout) {
+    // Block is deleted or is in a flyout.
+    return;
+  }
+  if (event.type == Blockly.Events.BLOCK_CREATE && event.ids.indexOf(this.id) != -1) {
+    // Check whether there is a matching function definition for this caller.
+    var name = this.getName();
+    var def = Blockly.Functions.getDefinition(name, this.workspace);
+    if (def) {
+      // The function definition exists, ensure the signatures match.
+      var defArgs = def.getArguments().slice();
+      defArgs.sort(function(a, b) { return a.name.localeCompare(b.name); });
```

Comment:
Wouldn't you want to not sort them if the signatures need to match?

Reply:
Each parameter has a UUID associated with it that is generated when it is added to the function. Since we do not support parameter reordering in the function editor, `func(foo: string, bar: bool)` and `func(bar: bool, foo: string)` will produce different serialized arguments, because either `bar` or `foo` will now have a new UUID (depending on which one was deleted + re-added to achieve the reordering).

Reply:
Hmm I guess there's no harm in not sorting the arrays. It'll be more future-proof in case we add parameter reordering in the function editor

Reply:
Ah, I see. I just meant that signatures are order dependent, e.g. 
`func(foo: number, bar: boolean) != func(bar: boolean, foo: number)` but I guess that isn't a real scenario

Reply:
There's actually no reason to sort, so I just got rid of it